cstrml/staking/src/lib.rs

```diff
@@ -1545,39 +1631,71 @@ impl<T: Trait> Module<T> {
         }
     }
 
-    /// Reward a given block author by a specific amount. Add reward to the block author's
-    fn reward_author(stash: &T::AccountId, reward: BalanceOf<T>) -> PositiveImbalanceOf<T> {
-          Self::make_payout(stash, reward).unwrap_or(<PositiveImbalanceOf<T>>::zero())
-    }
-
-    /// Reward a given (maybe)validator by a specific amount. Add the reward to the validator's, and its
-    /// guarantors' balance, pro-rata based on their exposure, after having removed the validator's
-    /// pre-payout cut.
-    fn reward_validator(stash: &T::AccountId, reward: BalanceOf<T>) -> PositiveImbalanceOf<T> {
-        // let reward = reward.saturating_sub(off_the_table);
-        let mut imbalance = <PositiveImbalanceOf<T>>::zero();
-        let validator_cut = if reward.is_zero() {
-            Zero::zero()
-        } else {
-            let exposure = Self::stakers(stash);
-            let total = exposure.total.max(One::one());
-            let total_rewards = Self::validators(stash).guarantee_fee * reward;
-            let mut guarantee_rewards = Zero::zero();
-
-            for i in &exposure.others {
-                let per_u64 = Perbill::from_rational_approximation(i.value, total);
-                // Reward guarantors
-                guarantee_rewards += per_u64 * total_rewards;
-                imbalance.maybe_subsume(Self::make_payout(&i.who, per_u64 * total_rewards));
-            }
+    fn do_reward_stakers(
+        validator_stash: T::AccountId,
+        era: EraIndex,
+    ) -> DispatchResult {
+        // Validate input data
+        let current_era = CurrentEra::get().ok_or(Error::<T>::InvalidEraToReward)?;
+        ensure!(era <= current_era, Error::<T>::InvalidEraToReward);
+        let history_depth = Self::history_depth();
+        ensure!(era >= current_era.saturating_sub(history_depth), Error::<T>::InvalidEraToReward);
+
+        // Note: if era has no reward to be claimed, era may be future. better not to update
+        // `ledger.claimed_rewards` in this case.
+        let era_staking_payout = <ErasStakingPayout<T>>::get(&era)
+            .ok_or_else(|| Error::<T>::InvalidEraToReward)?;
+
+        let controller = Self::bonded(&validator_stash).ok_or(Error::<T>::NotStash)?;
+        let mut ledger = <Ledger<T>>::get(&controller).ok_or_else(|| Error::<T>::NotController)?;
+
+        ledger.claimed_rewards.retain(|&x| x >= current_era.saturating_sub(history_depth));
```

Comment:
Personal question: Does thats means if the rewards are not claimed by v/g between `[current_era-history_depth, current_era]`, reward will be expired?😅

Reply:
Actuall this line is harmless. `claimed_rewards` is a field to avoid double reward. If the era is small than `current_era - history_depth`, it will be refused in line 1642. But the answer for your question is yes. Old information would be cleared.
```
        if let Some(old_era) = current_era.checked_sub(Self::history_depth() + 3) {
            Self::clear_era_information(old_era);
        }
```

Reply:
We `enable` it now, since it may be influent the economic mechanism, we can add `TODO` with ⚠️ emoji.

Reply:
How can we add emoji into code 😢 . I added one TODO without emoji